src/event/value.rs

```diff
@@ -18,6 +18,33 @@ pub enum Value {
     Null,
 }
 
+#[derive(PartialEq, Debug, Clone, Copy, is_enum_variant)]
+pub enum ValueKind {
```

Comment:
Interesting. Right now the remap language closely ties itself to `Value`, and that makes sense, since you want to be able to assign whatever return value you get to fields within the event.

Given that we want to still be able to provide field values as arguments to functions (e.g. `to_string(.foo, default = .bar)`), we need to stay in lock-step with `Value`.

Obviously the simplest solution is to accept a string value, and parse it as a regular expression, but that's fairly weakly typed, and requires us to validate the regular expression at runtime (which might be expensive).

I think the best course of action _for now_ is still to go this route, and then take [this comment I made earlier](https://github.com/timberio/vector/pull/3927#discussion_r492686564) into consideration for future improvements:

> I think this is a good place to start, but we can optionally expand function parameters to expose if they accept computed values or not.
> 
> Alternatively, functions can try to downcase a `dyn Function` to a `Literal`, in which case the resulting literal can be used without the runtime context, which allows the function to pre-compute whatever complex logic it needs, to improve performance.

Reply:
I do think there's value in being able to compute the regular expression (or any value) at runtime, so even if we had an explicit regular-expression value type for inputs, I'd still like it if you could supply that regular expression from another field, which kind of forces us to (also) support stringified regular expressions.

Reply:
We could of course update the PEG to parse `/<..>/` and have it be converted to `Value::Bytes`, that way you don't have to write them in quotation marks.

However, without some more plumbing work, that would only be an alias, so I do think we still want to be able to capture that type as being different from regular strings, and allow parameters to limit their input types to this regular expression type.

I think a possible solution would be:

We change `parameter.kinds` to be a list of enums, something like:

```rust
enum InputKind {
	Value(ValueKind),
	Regex,
}
```

A function would then (by choice) accept _both_ `InputKind::Value(ValueKind::Bytes)` _and_ `InputKind::Regex`. The former would compute the regex at runtime (but supports dynamic input from other fields), the latter could do the computation at compile-time.

There's probably other considerations to discuss, but this seems like a sensible way forward to allow specialized input values that don't map to `Value`, while still supporting dynamic field input.

Reply:
Yes that's true. But take, for example the [replace](https://github.com/timberio/vector/issues/3750) function. The parameter takes both a string and a regular expression. If you do supply the regex from another field, you would need a way of specifying if the field is a regex or a string. There are also additional flags that can be set on a regex - such as ignore case, which may not fit in a string.

What about if we created a special function that takes a string and returns a regex? So you could explicitly call it if you wanted a regex from a field:

```
.foo = replace(Regex(.field, ignorecase = true), "something")
```

The parser could parse a static regex string down to this. So `/a/i` would actually be `Regex("a", ignorecase = true)`.

Would there be a way to make that fit?

Reply:
> I think a possible solution would be:
> 
> We change `parameter.kinds` to be a list of enums, something like:
> 
> ```rust
> enum InputKind {
> 	Value(ValueKind),
> 	Regex,
> }
> ```
> 
> A function would then (by choice) accept _both_ `InputKind::Value(ValueKind::Bytes)` _and_ `InputKind::Regex`. The former would compute the regex at runtime (but supports dynamic input from other fields), the latter could do the computation at compile-time.
> 

That is similar to the way I implemented replace [here](https://github.com/FungusHumungus/vector/blob/remap_functions/src/mapping/query/functions.rs#L644).

Reply:
> There are also additional flags that can be set on a regex - such as ignore case, which may not fit in a string.

I'm not sure I follow. You could basically stick the entire regex in the string and parse that (e.g. `"/my regex/i"`). I agree it's less than ideal (and as I mentioned, weakly typed, since you'd basically check for `/`'s in the string to determine if its shape is like a regex), but it's doable.

> What about if we created a special function that takes a string and returns a regex?

This sounds interesting, but it does complicate things. It would probably be a `to_regex`, but obviously that function can't return a `Value` type, so now we have a split in function signatures, and we need to (somehow) convey which functions can be used where.

It's definitely a possibility, but it will increase cognitive complexity of the language.

Reply:
> That is similar to the way I implemented replace [here](https://github.com/FungusHumungus/vector/blob/remap_functions/src/mapping/query/functions.rs?rgh-link-date=2020-09-22T15%3A24%3A01Z#L644).

Indeed.

I think this will likely be the route we want to take. This would give us parser-level validation of regular expressions, while still keeping the 1-to-1 mapping with `Value`.

Reply:
>  You could basically stick the entire regex in the string and parse that (e.g. "/my regex/i").

Ah yes, that could work.

>  but it does complicate things.

It certainly does! That function could still return a Value. If it took a parameter of a closure - `Fn (Regex) -> Value` then anything that needed to use a regex would pass a closure to do the regex specific work it needs to do. The function would return a Value. There would need to be special code to handle it around the place, bit it could be an option if we want to stick with 'everything is a Value'.

I do still prefer the third option though as it does give us the tidiest option for further enhancements with other types down the line.

Reply:
> If it took a parameter of a closure - `Fn (Regex) -> Value` then anything that needed to use a regex would pass a closure to do the regex specific work it needs to do.

Right. My point was about this comment you made:

> What about if we created a special function that takes a string and returns a regex?

In that case, the function signature would have to be `Fn(String) -> Regex` (it would probably return an enum `Fn(String) -> Argument::Regex`), which we can't expose to the end-user similar to how we expose `to_string` or `md5`, etc.

So, this would be a special class of functions that can only be used as input to functions, you can't do `.foo = to_regex("/foobar/")`.

---

Regardless, I believe this discussion is moot, as we have a path forward that doesn't involve a function to convert a string to a regex.

We're going to update the Pest `argument` rule, something like:

```diff
argument = { positional_item | keyword_item }

- positional_item = { query_arithmetic }
+ positional_item = { argument_item }

- keyword_item = { ident ~ "=" ~ query_arithmetic }
+ keyword_item = { ident ~ "=" ~ argument_item }

+ argument_item = { query_arithmetic | regex }

+ regex = { "/" ~ /* TODO */ ~ "/" }
```

Then we'll change the parser to convert the input into an enum:

```rust
enum ArgumentKind {
	Value(event::Value),
	Regex(regex::Regex),
}
```

And finally, we update function signatures to have them either accept or reject `ArgumentKind::Regex`.

The only thing I'm still on the fence about is whether we should allow _dynamic regex_ by default on anything that accepts a regex, or if each function should build its own support for dynamic regexes, by accepting both `Value::Bytes` and `Regex` as an input value for a single parameter.

Basically, should we do something like this:

```rust
enum ArgumentKind {
	Value(event::Value),
	Regex(Regex),
}

enum Regex {
	Static(regex::Regex),
	Dynamic(bytes::Bytes),
}
```

And have the parser accept a string as regex input, and then have it compile the regex at runtime by implementing `Function` for `Regex` and call `regex::Regex::new` on the slice of bytes.

Reply:
Yes. That's a good approach.

I think regarding dynamic Regexes, there are functions that will accept both a String and a Regex. So, if the user is going to pass a dynamic Regex there will need to be a way for them to explicitly say that this field is a regex. Otherwise there will be no way to really know how to handle it.

eg.

```
.thing = replace(.field, .pattern, .with)
```

.pattern could be either a String or a Regex. It will be impossible to know which. There needs to be a way to disambiguate, such as:

```
.thing = replace(.field, regex(.pattern), .with)
```

Reply:
Yeah, that's right. I'm still not a fan of `regex()` though, as it still muddies the water of when a function can be used where. Again, because `.foo = regex(.pattern)` won't compile.

But I don't have a good alternative yet (other than inspecting the string to see if it matches`"/ ... /i..."`, so perhaps `regex()` _is_ the way forward, and we just need to document it as best as possible (or use other visual clues, such as `REGEX()` or `Regex` or `input_regex` or ... to distinguish input conversion functions from regular query functions.

Reply:
That is true.

Perhaps the problem is that we have a function that can accept both a regex or a string as a parameter. The problem would go away if we had seperate functions for each type of parameter - `replace` and `replace_regex`?

I think we should raise an issue for this and see if others have any ideas.

Reply:
Alternatively, we could have two separate optional arguments, one to replace using a regex, one using a string:

using `pattern` to use a static or dynamic regex:

* `replace(.foo, pattern = "/bar/i", with = "...")` — _must be a valid regex-formatted string_
* `replace(.foo, pattern = .bar, with = "...")` — _`.bar` must resolve to a valid regex-formatted string_
* `replace(.foo, pattern = /bar/i, with = "...")` — _compile-time checked to be a valid regex_

using `string` (or `literal`, or some other term) to use a static or dynamic string:

* `replace(.foo, string = "bar", with = "...")`
* `replace(.foo, string = .bar, with = "...")` — _`.bar` must resolve to a string_

Using both (`replace(.foo, pattern = .bar, string = .bar, with = "...")` would either use both to replace (with one occurring before the other) or result in a runtime error.